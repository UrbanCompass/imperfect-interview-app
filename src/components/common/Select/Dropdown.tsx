import React, { useContext, useMemo } from 'react';
import serialize from 'serialize-javascript';

import { SelectContext, SelectContextProps, Item } from './Select';
import { filterItems, filterOutSelectedItems } from './helpers';
import { Items } from './shared/Items';

export function Dropdown(): JSX.Element | null {
  const {
    // downshift
    inputValue,
    selectedItem,
    // props
    selectedItems = [],
    items,
    onlySelect,
    showAllItems,
  } = useContext(SelectContext) as SelectContextProps;

  if (selectedItem && selectedItems.length === 0) {
    selectedItems.push(selectedItem);
  }

  // match items by input
  const matchedItems = useMemo((): Item[] => {
    const rawMatch = filterItems({ items, inputValue });
    return filterOutSelectedItems({ items: rawMatch, selectedItems });
  }, [inputValue, serialize(items)]);

  // display only non-selected items (`onlySelect`)
  const remainingItems = useMemo((): Item[] => {
    return filterOutSelectedItems({ items, selectedItems });
  }, [serialize(items), serialize(selectedItems)]);

  const itemsToDisplay = showAllItems ? items : onlySelect ? remainingItems : matchedItems;

  return Array.isArray(itemsToDisplay) && itemsToDisplay.length > 0 ? (
    <Items itemsToDisplay={itemsToDisplay} />
  ) : null;
}
