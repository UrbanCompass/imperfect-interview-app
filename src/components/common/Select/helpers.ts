import toLower from 'lodash/toLower';
import { Item } from './Select';

export function itemToString(item): string {
  return item ? (item.value ? item.value : '') : '';
}

export function filterItems({ items, inputValue }): Item[] {
  return !inputValue
    ? items
    : items.filter(({ value }): boolean => toLower(value).includes(toLower(inputValue)));
}

export function filterOutSelectedItems({ items, selectedItems }): Item[] {
  return selectedItems.length === 0
    ? items
    : items.filter((item): boolean => {
        return !selectedItems.some((selected): boolean => {
          return item.id === selected.id;
        });
      });
}
