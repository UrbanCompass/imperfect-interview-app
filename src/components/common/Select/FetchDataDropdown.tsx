import React, { useContext, useMemo } from 'react';
import serialize from 'serialize-javascript';
import isEmpty from 'lodash/isEmpty';

import { useTypeaheadLoad } from '@/hooks/data/useTypeaheadLoad';
import { SelectContext, SelectContextProps, Item } from './Select';
import { Items } from './shared/Items';
import { filterOutSelectedItems } from './helpers';

export function FetchDataDropdown(): JSX.Element | null {
  const {
    inputValue,
    selectedItem,
    /* istanbul ignore next */
    fetchDataProps = { loadFn: {}, makeLoadProps: {}, transformToItems: {} },
    selectedItems = [],
  } = useContext(SelectContext) as SelectContextProps;

  if (selectedItem && selectedItems.length === 0) {
    selectedItems.push(selectedItem);
  }

  const { loadFn, makeLoadProps, transformToItems } = fetchDataProps;
  /* istanbul ignore next  */
  const { data = {}, loaded, ...rest } = useTypeaheadLoad({
    inputValue,
    loadFn,
    makeLoadProps,
  });

  const itemsToDisplay = useMemo((): Item[] => {
    if (loaded && !isEmpty(data) && typeof transformToItems === 'function') {
      const rawMatch = transformToItems(data);
      return filterOutSelectedItems({ items: rawMatch, selectedItems });
    }
    return [];
  }, [serialize(data)]);

  // always render so `Items` can display loading or loadError
  return inputValue ? <Items itemsToDisplay={itemsToDisplay} {...rest} /> : null;
}
