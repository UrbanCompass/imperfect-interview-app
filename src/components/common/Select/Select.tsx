import React, { createContext } from 'react';
import Downshift, { ControllerStateAndHelpers, StateChangeOptions } from 'downshift';
import isEmpty from 'lodash/isEmpty';

import { SelectInput } from './SelectInput';
import { Dropdown } from './Dropdown';
import { itemToString } from './helpers';
import { FetchDataDropdown } from './FetchDataDropdown';

export const SelectContext = createContext({});

export interface Item {
  value: string | number;
  id: string;
  style?: Record<string, string>;
}

interface SelectProps {
  autoFocus?: boolean;
  placeholder?: string;
  inputId: string;
  labelText?: string;
  items?: Item[];
  selectedItems?: Item[];
  selectedItem?: Item | undefined;
  onChange: (val: any, actions?: any) => void;
  onStateChange?: (
    options: StateChangeOptions<Item>,
    stateAndHelpers: ControllerStateAndHelpers<Item>
  ) => void;
  onInputValueChange?: (val: any, actions?: any) => void;
  stateReducer?: (state, changes) => Partial<StateChangeOptions<any>>;
  clear?(): void;
  removeItem?(item: Item): void;
  onlySelect?: boolean;
  highlightIndex?: number;
  initialInputValue?: string;
  fetchDataProps?: {
    loadFn: Function;
    makeLoadProps: Function;
    transformToItems: Function;
  };
  inputStyle?: Record<string, string>;
  containerStyle?: Record<string, string>;
  disableInput?: boolean;
  showAllItems?: boolean;
  inputIndex?: number;
  value?: string;
  onValueChange?(val: string): void;
  onEnterKey?: (val?: string | null) => void;
  onTabKey?: (val?: string | null) => void;
  initialIsOpen?: boolean;
  hideDropdown?: boolean;
}

export type SelectContextProps = SelectProps & ControllerStateAndHelpers<Item> & { id?: string };

function Select({
  onChange,
  onStateChange,
  onInputValueChange,
  labelText,
  inputId,
  fetchDataProps,
  highlightIndex,
  initialInputValue,
  disableInput,
  initialIsOpen,
  containerStyle,
  stateReducer,
  hideDropdown,
  ...rest
}: SelectProps): JSX.Element {
  return (
    <Downshift
      id={`downshift-${inputId}`}
      onChange={onChange}
      onStateChange={onStateChange}
      onInputValueChange={onInputValueChange}
      itemToString={itemToString}
      defaultHighlightedIndex={highlightIndex}
      initialInputValue={initialInputValue}
      initialIsOpen={initialIsOpen}
      stateReducer={stateReducer}
    >
      {(downshift): JSX.Element => {
        const { getLabelProps, isOpen } = downshift;

        return (
          <div>
            <SelectContext.Provider
              value={{
                ...downshift,
                ...rest,
                fetchDataProps,
                inputId,
                disableInput,
                containerStyle,
              }}
            >
              {/* eslint-disable-next-line jsx-a11y/label-has-for */}
              <label
                htmlFor={inputId}
                className="cx-formElement"
                {...getLabelProps({ css: { position: 'relative', ...containerStyle } })}
              >
                {labelText && <p className="cx-formElement-label">{labelText}</p>}
                <SelectInput />
                {hideDropdown
                  ? null
                  : isOpen && (!isEmpty(fetchDataProps) ? <FetchDataDropdown /> : <Dropdown />)}
              </label>
            </SelectContext.Provider>
          </div>
        );
      }}
    </Downshift>
  );
}

export { Select };
