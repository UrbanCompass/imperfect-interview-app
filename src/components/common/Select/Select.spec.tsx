import React from 'react';
import { shallow } from 'enzyme';
import { render as renderRoot, cleanup, fireEvent } from '@testing-library/react';

import { Select } from './Select';

describe('Select', (): void => {
  let getByPlaceholderText, getByRole, getByText;
  const onChange = jest.fn();
  const clear = jest.fn();
  const props = {
    items: [
      { value: 'foo', id: 'foo' },
      { value: 'bar', id: 'bar' },
      { value: 'foobar', id: 'foobar' },
    ],
    selectedItems: [],
    labelText: 'my label',
    onChange,
    clear,
    inputId: 'my-id',
    fakeTopLevelProp: true,
    placeholder: 'enter',
  };
  const downshiftProps = { getLabelProps: jest.fn(), fakeDownshiftProp: true };

  let wrapper = shallow(<Select {...props} />);

  beforeEach((): void => {
    ({ getByPlaceholderText, getByText, getByRole } = renderRoot(<Select {...props} />));
    onChange.mockClear();
  });

  afterEach(cleanup);

  it('renders Downshift', (): void => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders closed downshift content', (): void => {
    const render = wrapper.prop('children');
    const output = render({ ...downshiftProps, isOpen: false });
    expect(output).toMatchSnapshot();
  });

  it('renders open downshift content', (): void => {
    const render = wrapper.prop('children');
    const output = render({ ...downshiftProps, isOpen: true });
    expect(output).toMatchSnapshot();
  });

  it('renders no dropdown', (): void => {
    wrapper = shallow(<Select {...props} hideDropdown={true} />);
    const render = wrapper.prop('children');
    const output = render({ ...downshiftProps });
    expect(output).toMatchSnapshot();
  });

  it('handles change', (): void => {
    const downshift = wrapper.find('Downshift');
    const change = downshift.prop('onChange');
    change();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  it('displays dropdown on input', (): void => {
    fireEvent.change(getByPlaceholderText('enter'), { target: { value: 'a' } });
    expect(getByText('bar')).toBeTruthy();
    expect(getByText('foobar')).toBeTruthy();
  });

  it('set input on select item in dropdown', (): void => {
    fireEvent.change(getByPlaceholderText('enter'), { target: { value: 'a' } });
    fireEvent.click(getByText('foobar'));
    expect(onChange).toHaveBeenCalled();
  });

  it('click on x button would call clear', (): void => {
    fireEvent.change(getByPlaceholderText('enter'), { target: { value: 'a' } });
    fireEvent.click(getByText('foobar'));
    fireEvent.click(getByRole('button'));
    expect(clear).toHaveBeenCalled();
  });

  it('renders FetchedDropdown if options given, and dropdown Open', (): void => {
    const fetchDataProps = {
      loadFn: (): void => {},
      makeLoadProps: (): void => {},
      transformToItems: (): void => {},
    };
    const shallowWrapper = shallow(<Select {...props} fetchDataProps={fetchDataProps} />);
    const render = shallowWrapper.prop('children');
    const output = render({ ...downshiftProps, isOpen: true });
    expect(output).toMatchSnapshot();
  });
});
