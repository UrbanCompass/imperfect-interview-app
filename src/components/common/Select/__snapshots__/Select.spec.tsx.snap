// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders Downshift 1`] = `
<Downshift
  defaultHighlightedIndex={null}
  defaultIsOpen={false}
  environment={[Window]}
  getA11yStatusMessage={[Function]}
  id="downshift-my-id"
  itemToString={[Function]}
  onChange={[MockFunction]}
  onInputValueChange={[Function]}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  scrollIntoView={[Function]}
  selectedItemChanged={[Function]}
  stateReducer={[Function]}
  suppressRefError={false}
>
  <Component />
</Downshift>
`;

exports[`Select renders FetchedDropdown if options given, and dropdown Open 1`] = `
<div>
  <Context.Provider
    value={
      Object {
        "clear": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "containerStyle": undefined,
        "disableInput": undefined,
        "fakeDownshiftProp": true,
        "fakeTopLevelProp": true,
        "fetchDataProps": Object {
          "loadFn": [Function],
          "makeLoadProps": [Function],
          "transformToItems": [Function],
        },
        "getLabelProps": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "inputId": "my-id",
        "isOpen": true,
        "items": Array [
          Object {
            "id": "foo",
            "value": "foo",
          },
          Object {
            "id": "bar",
            "value": "bar",
          },
          Object {
            "id": "foobar",
            "value": "foobar",
          },
        ],
        "placeholder": "enter",
        "selectedItems": Array [],
      }
    }
  >
    <label
      className="cx-formElement"
      htmlFor="my-id"
    >
      <p
        className="cx-formElement-label"
      >
        my label
      </p>
      <SelectInput />
      <FetchDataDropdown />
    </label>
  </Context.Provider>
</div>
`;

exports[`Select renders closed downshift content 1`] = `
<div>
  <Context.Provider
    value={
      Object {
        "clear": [MockFunction],
        "containerStyle": undefined,
        "disableInput": undefined,
        "fakeDownshiftProp": true,
        "fakeTopLevelProp": true,
        "fetchDataProps": undefined,
        "getLabelProps": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "inputId": "my-id",
        "isOpen": false,
        "items": Array [
          Object {
            "id": "foo",
            "value": "foo",
          },
          Object {
            "id": "bar",
            "value": "bar",
          },
          Object {
            "id": "foobar",
            "value": "foobar",
          },
        ],
        "placeholder": "enter",
        "selectedItems": Array [],
      }
    }
  >
    <label
      className="cx-formElement"
      htmlFor="my-id"
    >
      <p
        className="cx-formElement-label"
      >
        my label
      </p>
      <SelectInput />
    </label>
  </Context.Provider>
</div>
`;

exports[`Select renders no dropdown 1`] = `
<div>
  <Context.Provider
    value={
      Object {
        "clear": [MockFunction],
        "containerStyle": undefined,
        "disableInput": undefined,
        "fakeDownshiftProp": true,
        "fakeTopLevelProp": true,
        "fetchDataProps": undefined,
        "getLabelProps": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "inputId": "my-id",
        "items": Array [
          Object {
            "id": "foo",
            "value": "foo",
          },
          Object {
            "id": "bar",
            "value": "bar",
          },
          Object {
            "id": "foobar",
            "value": "foobar",
          },
        ],
        "placeholder": "enter",
        "selectedItems": Array [],
      }
    }
  >
    <label
      className="cx-formElement"
      htmlFor="my-id"
    >
      <p
        className="cx-formElement-label"
      >
        my label
      </p>
      <SelectInput />
    </label>
  </Context.Provider>
</div>
`;

exports[`Select renders open downshift content 1`] = `
<div>
  <Context.Provider
    value={
      Object {
        "clear": [MockFunction],
        "containerStyle": undefined,
        "disableInput": undefined,
        "fakeDownshiftProp": true,
        "fakeTopLevelProp": true,
        "fetchDataProps": undefined,
        "getLabelProps": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
            Array [
              Object {
                "css": Object {
                  "position": "relative",
                },
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "inputId": "my-id",
        "isOpen": true,
        "items": Array [
          Object {
            "id": "foo",
            "value": "foo",
          },
          Object {
            "id": "bar",
            "value": "bar",
          },
          Object {
            "id": "foobar",
            "value": "foobar",
          },
        ],
        "placeholder": "enter",
        "selectedItems": Array [],
      }
    }
  >
    <label
      className="cx-formElement"
      htmlFor="my-id"
    >
      <p
        className="cx-formElement-label"
      >
        my label
      </p>
      <SelectInput />
      <Dropdown />
    </label>
  </Context.Provider>
</div>
`;
